src/accountsdb/index.zig

```diff
@@ -146,214 +170,170 @@ pub const AccountIndex = struct {
+        const bin, const bin_lg = self.getBinFromPubkey(pubkey).writeWithLock();
+        const ref_head = bin.get(pubkey.*) orelse return null;
+        return .{ ref_head, bin_lg };
+    }
+
+    pub const GetAccountRefError = error{ SlotNotFound, PubkeyNotFound };
+
+    /// Get a pointer to the account reference pointer with slot `slot` and pubkey `pubkey`,
+    /// alongside the write lock guard for the parent bin, and thus by extension the account
+    /// reference; this also locks access to all other account references in the parent bin.
+    /// This can be used to update an account reference (ie by replacing the `*AccountRef`).
+    pub fn getReferencePtrPtrWrite(
+        self: *const Self,
+        pubkey: *const Pubkey,
+        slot: Slot,
+    ) GetAccountRefError!struct { **AccountRef, RwMux(RefMap).WLockGuard } {
+        const bin, var bin_lg = self.getBinFromPubkey(pubkey).writeWithLock();
+        errdefer bin_lg.unlock();
+
+        const head_ref = bin.getPtr(pubkey.*) orelse return error.PubkeyNotFound;
+        const ref_ptr_ptr = switch (head_ref.getPtrToFieldThatIsPtrToRefWithSlot(slot)) {
+            .null => return error.SlotNotFound,
+            .head => &head_ref.ref_ptr,
+            .inner => |inner| &inner.*.?,
+        };
+        return .{ ref_ptr_ptr, bin_lg };
     }
 
     /// returns a reference to the slot in the index which is a local copy
     /// useful for reading the slot without holding the lock.
     /// NOTE: its not safe to read the underlying data without holding the lock
-    pub fn getReferenceSlot(self: *Self, pubkey: *const Pubkey, slot: Slot) ?AccountRef {
-        var head_ref_rw = self.getReference(pubkey) orelse return null;
-        const head_ref, var head_ref_lg = head_ref_rw.readWithLock();
+    pub fn getReferenceSlotCopy(self: *Self, pubkey: *const Pubkey, slot: Slot) ?AccountRef {
+        const head_ref, var head_ref_lg = self.getReferenceHeadRead(pubkey) orelse return null;
         defer head_ref_lg.unlock();
 
         var curr_ref: ?*AccountRef = head_ref.ref_ptr;
-        const slot_ref = while (curr_ref) |ref| : (curr_ref = ref.next_ptr) {
+        var slot_ref_copy: AccountRef = while (curr_ref) |ref| : (curr_ref = ref.next_ptr) {
             if (ref.slot == slot) break ref.*;
-        } else null;
-
-        return slot_ref;
+        } else return null;
+        // since this will purely be a copy, it's safer to not allow the caller
+        // to observe the `next_ptr` value, because they won't have the lock.
+        slot_ref_copy.next_ptr = null;
+        return slot_ref_copy;
     }
 
     pub fn exists(self: *Self, pubkey: *const Pubkey, slot: Slot) bool {
-        var head_reference_rw = self.getReference(pubkey) orelse return false;
-        const head_ref, var head_reference_lg = head_reference_rw.readWithLock();
-        defer head_reference_lg.unlock();
+        const head_ref, var head_ref_lg = self.getReferenceHeadRead(pubkey) orelse return false;
```

Comment:
should be bin_lg

Reply:
a4672c0d5bc5605471eb012f128b133d61d2ec99